extensions/kafka-eight-simpleConsumer/src/main/java/io/druid/firehose/kafka/KafkaSimpleConsumer.java

```diff
@@ -0,0 +1,405 @@
+  }
+
+  private void stopConsumer()
+  {
+    if (consumer != null) {
+      try {
+        consumer.close();
+        log.info("stop consumer[%s][%s], leaderBroker[%s]", topic, partitionId, leaderBroker);
+      }
+      catch (Exception e) {
+        log.warn(e, "stop consumer[%s][%s] failed", topic, partitionId);
+      }
+      finally {
+        consumer = null;
+      }
+    }
+  }
+
+  public void stop()
+  {
+    stopConsumer();
+    log.info("KafkaSimpleConsumer[%s][%s] stopped", topic, partitionId);
+  }
+
+  private PartitionMetadata findLeader() throws InterruptedException
+  {
+    for (KafkaBroker broker : replicaBrokers) {
+      SimpleConsumer consumer = null;
+      try {
+        log.info("Finding new leader from Kafka brokers, try broker %s:%s", broker.host, broker.port);
+        consumer = new SimpleConsumer(broker.host, broker.port, SO_TIMEOUT, BUFFER_SIZE, leaderLookupClientId);
+        TopicMetadataResponse resp = consumer.send(new TopicMetadataRequest(Collections.singletonList(topic)));
+
+        List<TopicMetadata> metaData = resp.topicsMetadata();
+        for (TopicMetadata item : metaData) {
+          if (topic.equals(item.topic())) {
+            for (PartitionMetadata part : item.partitionsMetadata()) {
+              if (part.partitionId() == partitionId) {
+                return part;
+              }
+            }
+          }
+        }
+      }
+      catch (Exception e) {
+        ensureNotInterrupted(e);
+        log.warn(
+            e, "error communicating with Kafka Broker [%s] to find leader for [%s] - [%s]", broker, topic, partitionId
+        );
+      }
+      finally {
+        if (consumer != null) {
+          consumer.close();
+        }
+      }
+    }
+
+    return null;
+  }
+
+  private Broker findNewLeader(Broker oldLeader) throws InterruptedException
+  {
+    long retryCnt = 0;
+    while (true) {
+      PartitionMetadata metadata = findLeader();
+      if (metadata != null) {
+        replicaBrokers.clear();
+        for (Broker replica : metadata.replicas()) {
+          replicaBrokers.add(new KafkaBroker(replica.host(), replica.port()));
+        }
+
+        log.debug("Got new Kafka leader metadata : [%s], previous leader : [%s]", metadata, oldLeader);
+        Broker newLeader = metadata.leader();
+        if (newLeader != null) {
+          // We check the retryCnt here as well to make sure that we have slept a little bit
+          // if we don't notice a change in leadership
+          // just in case if Zookeeper doesn't get updated fast enough
+          if (oldLeader == null || isValidNewLeader(newLeader) || retryCnt != 0) {
+            return newLeader;
+          }
+        }
+      }
+
+      Thread.sleep(RETRY_INTERVAL);
+      retryCnt++;
+      // if could not find the leader for current replicaBrokers, let's try to
+      // find one via allBrokers
+      if (retryCnt >= 3 && (retryCnt - 3) % 5 == 0) {
```

Comment:
This is just to throttle the "reset all brokers" behavior and error log so that it won't show up each time 'cannot find leader' happens. I would totally agree that this should be changed to some wrapped retry mechanism, or even removing this part won't impact functionality.